Add test-case annotation. (#2764)
modelica/ModelicaSpecification: chapters/annotations.tex

\begin{example}
\begin{lstlisting}[language=modelica]
annotation (

>    Icon(coordinateSystem(extent = {{-100, -100}, {100, 100}}),<

Reviewer: Typo?
```suggestion
   Icon(coordinateSystem(extent = {{-100, -100}, {100, 100}}),
```

\end{semantics}
\end{annotationdefinition}

\section{Annotations for Simulations}\label{annotations-for-simulations}

> These annotations define how models can be simulated.

Reviewer: Did we say that a test case could be more general than something that would be possible to simulate even if errors were corrected?  For example, couldn't a partial model be a test case that demonstrates some illegal use?

Author: I changed that to be more general.

Reviewer: Yes, the new introduction is a better description of the two subsections, but maybe the two annotations are simply too different to fit under the same section?

Author: I still think they are related. However, I think we should more do a pass to make that chapter more consistent; as sections vary wildly in length.

Reviewer: OK, seems like we've had enough of discussion for now then.

The \lstinline!experiment! annotation defines the default start time (\lstinline!StartTime!) in {[}s{]}, the default stop time (\lstinline!StopTime!) in {[}s{]}, the suitable time resolution for the result grid (\lstinline!Interval!) in {[}s{]}, and the default relative integration tolerance (\lstinline!Tolerance!) for simulation experiments to be carried out with the model or block at hand.  If \lstinline!StartTime! is not specified it is assumed to be \lstinline!0.0!.

\subsection{Annotation for Test Cases}\label{annotation-for-test-cases}

> \begin{lstlisting}[language=grammar]

Reviewer: Another way to structure _Annotations for Simulations_ would be to use two `annotationdefinition` environments instead of two subsections.  This way, the new annotation would automatically end up in the index.

Reviewer: Resolving with the same resolution as the comment above.

testcase-annotation:
   annotation "(" "TestCase" "(" "shouldPass" "=" ( false | true ) ")" ")"
\end{lstlisting}
If \lstinline!shouldPass! is \lstinline!false! it indicates that the translation or the simulation of the model should fail.

> If a tools checks a package where classes have \lstinline!shouldPass=false! they should not generate errors - and testing them may even be skipped; if a model with \lstinline!shouldPass=false! translates and simulates without error that should generate a diagnostics.

Reviewer: I find the dash before _and_ a bit odd, how about a comma?  Perhaps also split into two sentences?

I also find the diagnostics part maybe a bit too challenging for tools to handle, and the information in the annotation insufficient to directly result in tests of much value.  (Separation into `shouldTranslate` and `shouldSimulate` would have resulted in somewhat better tests.)  Isn't it better to say more generally that the model may be useful for purposes of negative testing?

```suggestion
If a tools checks a package where classes have \lstinline!shouldPass=false! they should not generate errors, and may even be skipped.
On the other hand, models with \lstinline!shouldPass=false! may be useful for creation of negative tests in tool-specific ways.
```

Author: Good suggestion - but I will improve 'may even be skipped' to include that it's checking that is skipped.

> Similarly as a class with obsolete-annotation a class with \lstinline!TestCase! annotation (regardless of the value of \lstinline!shouldPass!) shall not be used in other models, unless those models also have a \lstinline!TestCase! annotation.

Reviewer: That famous comma:
```suggestion
Similarly as a class with obsolete-annotation, a class with \lstinline!TestCase! annotation (regardless of the value of \lstinline!shouldPass!) shall not be used in other models, unless those models also have a \lstinline!TestCase! annotation.
```
